Converts Dispatcher().RunAsync to WinRT Coroutines

mkitzan opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request
This PR turns all* instances of `Dispatcher().RunAsync` to WinRT coroutines 👌. 
This was good coding fodder to fill my plane ride ✈️. Enjoy your holidays everyone!

*With the exception of three functions whose signatures cannot be changed due to inheritance and function overriding in `TermControlAutomationPeer` [`L44`](https://github.com/microsoft/terminal/blob/master/src/cascadia/TerminalControl/TermControlAutomationPeer.cpp#L44), [`L58`](https://github.com/microsoft/terminal/blob/master/src/cascadia/TerminalControl/TermControlAutomationPeer.cpp#L58),  [`L72`](https://github.com/microsoft/terminal/blob/master/src/cascadia/TerminalControl/TermControlAutomationPeer.cpp#L72). 

<!-- Other than the issue solved, is this relevant to any other issues/existing PRs? --> 
## References

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist
* [x] Closes #3919
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [ ] Tests added/passed
* [ ] Requires documentation to be updated
* [x] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #3919

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments
My thought pattern here was to minimally disturb the existing code where possible. So where I could, I converted existing functions into coroutine using functions (like in the [core example](https://github.com/microsoft/terminal/issues/3919#issue-536598706)). For ~the most part~ all instances, I used the format where [`this` is accessed safely within a locked scope](https://github.com/microsoft/terminal/issues/3919#issuecomment-564730620). Some function signatures were changed to take objects by value instead of reference, so the coroutines don't crash when the objects are accessed past their original lifetime. The [copy](https://github.com/microsoft/terminal/blob/master/src/cascadia/TerminalApp/TerminalPage.cpp#L1132) and [paste](https://github.com/microsoft/terminal/blob/master/src/cascadia/TerminalApp/TerminalPage.cpp#L1170) event handler entry points were originally set to a high priority; however, the WinRT coroutines don't appear to support a priority scheme so this priority setting was not preserved in the translation.

<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed
Compiles and runs, and for every event with a clear trigger repro, I triggered it to ensure crashes weren't introduced.

DHowett-MSFT commented:
@mkitzan you’re a star. I’ll put this on my review queue but can’t promise much over the holidays :) enjoy your travel!

mkitzan commented:
No worries @DHowett-MSFT, I'll be too sloshed this week in the prairies to make any meaningful revisions anyways. Enjoy your holidays!

mkitzan commented:
Going to do a pass on these files to convert every handler to use `this` within a locked scope.

miniksa commented on src/cascadia/TerminalApp/Tab.cpp:
```diff
@@ -317,13 +318,13 @@ void Tab::_AttachEventHandlersToControl(const TermControl& control)
     std::weak_ptr<Tab> weakThis{ shared_from_this() };
 
-    control.TitleChanged([weakThis](auto newTitle) {
+    control.TitleChanged([weakThis, this](auto newTitle) {
```
Why did you also capture `this` instead of just using the `tab`? (Here and below).

zadjii-msft replied:
I agree, I think we should probably _not_ capture `this` here at all, to make sure we don't accidentally use that instead of the `weakThis.lock()` result

miniksa commented on src/cascadia/TerminalApp/TerminalPage.cpp:
```diff
@@ -63,33 +63,30 @@ namespace winrt::TerminalApp::implementation
-        auto weakThis{ get_weak() };
-
-        _tabView.TabDragStarting([weakThis](auto&& /*o*/, auto&& /*a*/) {
+        _tabView.TabDragStarting([weakThis{ get_weak() }, this](auto&& /*o*/, auto&& /*a*/) {
```
This file too... I feel better about using the actual strong pointer than by wink-nudge using the `this` simply because it turned out that the weak could be strengthened.

miniksa commented on src/cascadia/TerminalControl/TermControl.cpp:
```diff
@@ -533,24 +535,41 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
     }
 
-    void TermControl::SwapChainChanged()
+    winrt::fire_and_forget TermControl::SwapChainChanged()
```
While reading some background documentation on how this whole co* stuff worked, I noticed that `SwapChainPanel` apparently has a provision to let background threads potentially mess with the chain instead of having to jump to the UI thread to do so:

https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.swapchainpanel.createcoreindependentinputsource#Windows_UI_Xaml_Controls_SwapChainPanel_CreateCoreIndependentInputSource_Windows_UI_Core_CoreInputDeviceTypes_

It's probably not easily/instantly applicable here. But I want to leave the note because I thought of it and @DHowett-MSFT and @zadjii-msft might have an opinion on whether that would be a decent follow-on task to generate from this comment.

miniksa replied:
Moved to #4164 for further investigation.

mkitzan commented:
@miniksa, I'll add a spot in the docs highlighting the C++/WinRT concurrency and weak pointers information.

As for the implicit `this` usage, I decided to go with it because it was the pattern accepted in #3908. Despite that PR being accepted, there was no real decision on the pattern (the PR was accepted w/out further discussion). There have been three main patterns suggested:
1. Using the strengthened `strongThis` ([concerns exist](https://github.com/microsoft/terminal/pull/3908#discussion_r356857885) over the constant calls to `operator->()`)
2. Using implicit `this` within a safe scope ([concerns exist](https://github.com/microsoft/terminal/issues/3919#issuecomment-564729365) over the ease to which mistakes may be introduced)
3. Using a reference to the object obtained by dereferencing `strongThis` [`auto& that{ *strongThis };`](https://github.com/microsoft/terminal/pull/3908#discussion_r357379989)

Patterns 1 and 2 both have compelling reasons for their adoption: avoiding programmer error, and performance respectively. Pattern 3 hasn't been discussed much, but it seems like an interesting option though it requires remembering to dereference the `strongThis`. Also, the majority of the lambdas are pretty short making the [readability benefits](https://github.com/microsoft/terminal/pull/3908#discussion_r356887102) I found mostly meaningless.

Whatever pattern is decided, I'll make sure it's implemented uniformly across the code base.

miniksa commented:
> @miniksa Michael Niksa FTE, I'll add a spot in the docs highlighting the C++/WinRT concurrency and weak pointers information.
> 
> As for the implicit `this` usage, I decided to go with it because it was the pattern accepted in #3908. Despite that PR being accepted, there was no real decision on the pattern (the PR was accepted w/out further discussion). There have been three main patterns suggested:
> 
> 1. Using the strengthened `strongThis` ([concerns exist](https://github.com/microsoft/terminal/pull/3908#discussion_r356857885) over the constant calls to `operator->()`)
> 2. Using implicit `this` within a safe scope ([concerns exist](https://github.com/microsoft/terminal/issues/3919#issuecomment-564729365) over the ease to which mistakes may be introduced)
> 3. Using a reference to the object obtained by dereferencing `strongThis` [`auto& that{ *strongThis };`](https://github.com/microsoft/terminal/pull/3908#discussion_r357379989)
> 
> Patterns 1 and 2 both have compelling reasons for their adoption: avoiding programmer error, and performance respectively. Pattern 3 hasn't been discussed much, but it seems like an interesting option though it requires remembering to dereference the `strongThis`. Also, the majority of the lambdas are pretty short making the [readability benefits](https://github.com/microsoft/terminal/pull/3908#discussion_r356887102) I found mostly meaningless.
> 
> Whatever pattern is decided, I'll make sure it's implemented uniformly across the code base.

Good analysis. I don't think I can make that decision in a vacuum, though. And I don't think I saw that previous discussion thread where @DHowett-MSFT and @zadjii-msft chimed in with their opinions.

It looks like I was saying I was in favor of 1, @zadjii-msft based on that discussion preferred 1, and @DHowett-MSFT preferred 2 but proposed 3.

I personally think programmer error is the most likely scenario for asynchronous behavior, so that's why I'd like to bias toward eliminating that over performance until proven that the performance benefit is needed.

However, I'd like to see @DHowett-MSFT and @zadjii-msft chime in one final time here and make a formal decision together. That might require waiting until next week after the holidays unless they're both checking mail.

DHowett-MSFT commented:
Alright, so... I won't die on the "`strongThis.operator->()` is bad" hill. I'm sure a smart compiler can determine that the pointer living in `winrt::com_ptr` doesn't change. :smile:

@mkitzan I wanted to chip in on this, though:

> the WinRT coroutines don't appear to support a priority scheme

I've got some real excellent news (and this comes at my own surprise!)

C++/WinRT supports `co_await winrt::resume_foreground(queue, DispatcherQueuePriority::High);` :grin:

(sorry: still holidaying - I'm here to comment but not yet review)

mkitzan commented:
That coroutine priority info is great. I've added back the high priority for copy-paste coroutines, and added the C++/WinRT concurrency links into the documentation.

zadjii-msft commented on src/cascadia/TerminalApp/Pane.cpp:
```diff
@@ -730,6 +730,13 @@ void Pane::_CloseChild(const bool closeFirst)
 
+winrt::fire_and_forget Pane::_CloseChildRoutine(const bool closeFirst)
+{
+    co_await winrt::resume_foreground(_root.Dispatcher());
```
I'm mildly worried that this particular callback also falls under the category of things where we should grab a weak ref to `this` at the top, then get the strong ref after the `resume_foreground` call.

mkitzan replied:
Good catch, the previous lambdas were capturing by copy

zadjii-msft commented on src/cascadia/TerminalApp/TerminalPage.cpp:
```diff
@@ -882,18 +877,19 @@ namespace winrt::TerminalApp::implementation
     {
         // GH#1117: This is a workaround because _tabView.SelectedIndex(tabIndex)
         //          sometimes set focus to an incorrect tab after removing some tabs
         auto tab = _tabs.at(tabIndex);
```
Should this line be after the `co_await`?

mkitzan replied:
`auto tab = _tabs.at(tabIndex);` appears before the `RunAsync` call which is why it's before the `co_await`. Let me know if there's something I'm not seeing here / rationale I'm missing.

zadjii-msft replied:
This was mostly a thought experiment, it seems fine in either place tbh

zadjii-msft commented on src/cascadia/TerminalApp/TerminalPage.cpp:
```diff
@@ -1165,12 +1161,12 @@ namespace winrt::TerminalApp::implementation
+    winrt::fire_and_forget TerminalPage::_PasteFromClipboardHandler(const IInspectable /*sender*/,
+                                                                    const PasteFromClipboardEventArgs eventArgs)
     {
-        this->Dispatcher().RunAsync(CoreDispatcherPriority::High, [eventArgs]() {
```
related? maybe us capturing the `PasteFromClipboardEventArgs&` here has something to do with #4118, and capturing the strong ref instead will help?

mkitzan replied:
It's hard to say if it'll fix #4118, but it's worth a shot.

DHowett-MSFT replied:
wow actually yeah that's totally reasonable

mkitzan commented:
@zadjii-msft thanks for the review. I'll have a commit to fix these later today, and it'll include the fix to use the strengthened weak_ptr.

mkitzan commented on src/cascadia/TerminalControl/TermControl.cpp:
```diff
@@ -1614,25 +1623,25 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
-        _lastScrollOffset = { viewTop };
-        _scrollPositionChangedHandlers(viewTop, viewHeight, bufferSize);
+                // Set this value as our next expected scroll position.
+                _lastScrollOffset = { viewTop };
```
Good catch, I'll move those lines

mkitzan commented:
The latest commit addresses the issues found in @zadjii-msft review, and make changes the lambda patterns to the agreed format. Also, there seems to be an azp hang.

zadjii-msft commented:
/azp run

DHowett-MSFT commented on doc/STYLE.md:
```diff
@@ -5,3 +5,4 @@
 1. If it's brand new code or refactoring a complete class or area of the code, please follow as Modern C++ of a style as you can and reference the [C++ Core Guidelines](https://github.com/isocpp/CppCoreGuidelines) as much as you possibly can.
 1. When working with any Win32 or NT API, please try to use the [Windows Implementation Library](./WIL.md) smart pointers and result handlers.
 1. The use of NTSTATUS as a result code is discouraged, HRESULT or exceptions are preferred. Functions should not return a status code if they would always return a successful status code. Any function that returns a status code should be marked `noexcept` and have the `nodiscard` attribute.
+1. When contributing code in `TerminalApp`, be mindful to appropriately use C++/WinRT [strong and weak references](https://docs.microsoft.com/en-us/windows/uwp/cpp-and-winrt-apis/weak-references), and have a good understanding of C++/WinRT [concurrency schemes](https://docs.microsoft.com/en-us/windows/uwp/cpp-and-winrt-apis/concurrency).
```
This is my favorite part. Thank you

DHowett-MSFT commented:
Aww, there's merge conflicts. That's too bad!

mkitzan commented:
Made a quick fix for the merge conflicts. Should be good to go 👌

DHowett-MSFT commented:
Yess. Thanks!

DHowett-MSFT commented:
@msftbot merge this in 1 hour
